src/Npgsql.DependencyInjection/NpgsqlServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Data.Common;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Logging;
+using Npgsql;
+
+// ReSharper disable once CheckNamespace
+namespace Microsoft.Extensions.DependencyInjection;
+
+/// <summary>
+/// Extension method for setting up Npgsql services in an <see cref="IServiceCollection" />.
+/// </summary>
+public static class NpgsqlServiceCollectionExtensions
+{
+    /// <summary>
+    /// Registers an <see cref="NpgsqlDataSource" /> and an <see cref="NpgsqlConnection" /> in the <see cref="IServiceCollection" />,
+    /// </summary>
+    /// <param name="serviceCollection">The <see cref="IServiceCollection" /> to add services to.</param>
+    /// <param name="connectionString">An Npgsql connection string.</param>
+    /// <param name="dataSourceBuilderAction">
+    /// An action to configure the <see cref="NpgsqlDataSourceBuilder" /> for further customizations of the <see cref="NpgsqlDataSource" />.
+    /// </param>
+    /// <param name="connectionLifetime">
+    /// The lifetime with which to register the <see cref="NpgsqlConnection" /> in the container.
+    /// Defaults to <see cref="ServiceLifetime.Scoped" />.
+    /// </param>
+    /// <param name="dataSourceLifetime">
+    /// The lifetime with which to register the <see cref="NpgsqlDataSource" /> service in the container.
+    /// Defaults to <see cref="ServiceLifetime.Singleton" />.
+    /// </param>
+    /// <returns>The same service collection so that multiple calls can be chained.</returns>
+    public static IServiceCollection AddNpgsqlDataSource(
+        this IServiceCollection serviceCollection,
+        string connectionString,
+        Action<NpgsqlDataSourceBuilder> dataSourceBuilderAction,
+        ServiceLifetime connectionLifetime = ServiceLifetime.Scoped,
+        ServiceLifetime dataSourceLifetime = ServiceLifetime.Singleton)
```

Comment:
It's worth keeping in mind that disposable transient services are not disposed until the container scope is disposed. So all transient services resolved from a scope will remain un-disposed until the scope is disposed, and transient services resolved from the root service provider will not be disposed until the root is disposed, which often means they won't be disposed until the application shuts down.

Another way of putting this is that it is a bad idea to resolve disposable transient services from the root provider.

Reply:
Was not aware of this, thanks... Important stuff.

The thing with DbConnection is that we can't inject an already-open connection, since opening as an async operation and DI doesn't support that... So users will have to call Open in their in code in any case, so it seems reasonable to require them to also close/dispose it. On a connection is closed, it's no longer important to also dispose it (it's just a memory object at that point, at least in Npgsql), so maybe that's a reasonable state of affairs?

It would mean that [this sample](https://github.com/npgsql/npgsql/pull/4521/files#diff-a03d4ba78b33b9e8db9381e6b03dbe780e6924f868f1024a26869582c6879e3aR18) currently leaks; either we add a using there, or we make the connection Scoped by default instead of Transient.

What do you think @ajcvickers? I agree that there's a pit of failure with Transient, where people will assume the connection gets closed/disposed by DI, and so they'll have a connection leak... Should we switch back to Scoped on this?

Reply:
> t's just a memory object at that point, at least in Npgsql

A connection object could have a native resource associated with it. For example, Sqlite without connection pooling, maybe?

> normal app

What is this "normal app" of which you speak?  :-P

Maybe you should call me and we can chat about this?

Reply:
>  So users will have to call Open in their in code in any case, so it seems reasonable to require them to also close/dispose it.

Typically, this points towards returning a factory rather than the object itself.

Reply:
Discussed this, and we think it's fine to register NpgsqlConnection as Transient:

* Normal ASP.NET applications aren't directly exposed to the root SP, so everything is fine - transient services resolved from the scoped SP get properly disposed when the scope is disposed.
* It's basically an error to resolve either Transient or Scoped from the root SP:
  * In the Transient case, the service doesn't get disposed until the root SP is disposed, so the user would be responsible for disposing. Even worse, the root SP keeps track of each Transient service it hands out (so it can be disposed at the end), so there's a memory leak
  * In the Scoped case, a singleton is returned instead (though exception in dev), which would mean a single global connection (which again doesn't get disposed until the root SP is disposed). That's almost always the wrong thing.
* Transient does have advantages over Scoped here:
  * Transient would allow concurrent usage in the same Scope (different connection instances)
  * There's no state on NpgsqlConnection that people typically want to share (unlike with DbContext)
  * Even without concurrent usage, better to not leak anything by default between different usage in the scope (e.g. one consumer opens but doesn't close, the other gets an already-open connection)
* So since the root SP scenario is wrong in any case and Transient is more appropriate, we'll register as Transient by default and guide people to avoid ever resolving connections on the root SP.